h2o-k8s/src/main/java/water/k8s/lookup/KubernetesDnsLookup.java

```diff
@@ -0,0 +1,164 @@
+ *           env:
+ *           - name: H2O_KUBERNETES_SERVICE_DNS
+ *             value: h2o-service.h2o-statefulset.svc.cluster.local
+ *           - name: H2O_NODE_LOOKUP_TIMEOUT
+ *             value: '180'
+ *           - name: H2O_NODE_EXPECTED_COUNT
+ *             value: '3'
+ * </pre>
+ */
+public class KubernetesDnsLookup implements KubernetesLookup {
+
+    private static final String K8S_SERVICE_DNS_ENV_VAR_KEY = "H2O_KUBERNETES_SERVICE_DNS";
+    private static final String DNS_TIMEOUT_DEFAULT = "30000"; // 30 seconds
+    private static final int ONE_SECOND = 1000;
+    private final String serviceDns;
+    private final DirContext dirContext;
+
+    public KubernetesDnsLookup(final String serviceDns) {
+        this.serviceDns = serviceDns;
+        this.dirContext = initDirContext();
+    }
+
+    /**
+     * @return
+     * @throws IllegalStateException When the H2O-related kubernetes DNS service is not found
+     */
+    public static KubernetesDnsLookup fromH2ODefaults() throws IllegalStateException {
+        final String dnsServiceName = System.getenv(K8S_SERVICE_DNS_ENV_VAR_KEY);
+        if (dnsServiceName == null) {
+            throw new IllegalStateException(String.format("DNS of H2O service not set. Please set the '%s' variable.",
+                    K8S_SERVICE_DNS_ENV_VAR_KEY));
+        } else if (dnsServiceName.trim().isEmpty()) {
+            throw new IllegalStateException(String.format("DNS Service '%s' name is invalid.", dnsServiceName));
+        }
+        return new KubernetesDnsLookup(dnsServiceName);
+    }
+
+    private static String extractHost(final String server, final Pattern extractHostPattern) {
+        String host = server.split(" ")[3];
+        return extractHostPattern.matcher(host).replaceAll("");
+    }
+
+    /**
+     * Looks up H2O pods via configured K8S Stateless service DNS. Environment variable with key defined in
+     * H2O_K8S_SERVICE_DNS_KEY constant is used to obtain address of the DNS. The DNS is then queried for pods
+     * in the underlying service. It is the responsibility of the K8S cluster owner to set-up the service correctly to
+     * only provide correct adresses of pods with H2O active. If pods with no H2O running are supplied, the resulting
+     * flatfile may contain pod IPs with no H2O running as well.
+     *
+     * @param lookupConstraints Constraints to obey during lookup
+     * @return A {@link Set} of adresses of looked up nodes represented as String. The resulting set is never empty.
+     */
+    public Optional<Set<String>> lookupNodes(final Collection<LookupConstraint> lookupConstraints) {
+        final Set<String> lookedUpNodes = new HashSet<>();
+
+        while (lookupConstraints.stream().allMatch(lookupStrategy -> !lookupStrategy.isLookupEnded(lookedUpNodes))) {
+            try {
+                dnsLookup(lookedUpNodes);
+                Thread.sleep(ONE_SECOND);
+            } catch (NamingException e) {
+                continue;
```

Comment:
can we produce a warning?

Reply:
Yes we can ! Done.

Beware: This might happen if the headless service is started at the same time as the H2O pods - During the very first iteration or a couple more iterations, this might actually happen, as the records are not there yet.